predicators/utils.py

```diff
@@ -1745,41 +1754,100 @@ def query(self, pt1: _BiRRTState,
-    def _rrt_connect(self, pt1: _BiRRTState,
-                     pt2: _BiRRTState) -> Optional[List[_BiRRTState]]:
-        root1, root2 = _BiRRTNode(pt1), _BiRRTNode(pt2)
-        nodes1, nodes2 = [root1], [root2]
+    def _rrt_connect(self, pt1: _RRTState,
+                     pt2: _RRTState) -> Optional[List[_RRTState]]:
+        root = _RRTNode(pt1)
+        nodes = [root]
+
+        for _ in range(self._num_iters):
+            # Sample the goal with a small probability, otherwise randomly
+            # choose a point.
+            sample_goal = self._rng.random() < self._sample_goal_eps
+            samp = pt2 if sample_goal else self._sample_fn(pt1)
+            min_key = functools.partial(self._get_pt_dist_to_node, samp)
+            nearest = min(nodes, key=min_key)
+            reached_goal = False
+            for newpt in self._extend_fn(nearest.data, samp):
+                if self._collision_fn(newpt):
+                    break
+                nearest = _RRTNode(newpt, parent=nearest)
+                nodes.append(nearest)
+            else:
+                reached_goal = sample_goal
+            # Check goal_fn if defined
+            if reached_goal or self._goal_fn is not None and self._goal_fn(
+                    nearest.data):
+                path = nearest.path_from_root()
+                return [node.data for node in path]
+        return None
 
-        def _get_pt_dist_to_node(pt: _BiRRTState,
-                                 node: _BiRRTNode[_BiRRTState]) -> float:
-            return self._distance_fn(pt, node.data)
+    def _get_pt_dist_to_node(self, pt: _RRTState,
+                             node: _RRTNode[_RRTState]) -> float:
+        return self._distance_fn(pt, node.data)
+
+    def _smooth_path(self, path: List[_RRTState]) -> List[_RRTState]:
+        assert len(path) > 2
+        for _ in range(self._smooth_amt):
+            i = self._rng.integers(0, len(path) - 1)
+            j = self._rng.integers(0, len(path) - 1)
+            if abs(i - j) <= 1:
+                continue
+            if j < i:
+                i, j = j, i
+            shortcut = list(self._extend_fn(path[i], path[j]))
+            if len(shortcut) < j-i and \
+               all(not self._collision_fn(pt) for pt in shortcut):
+                path = path[:i + 1] + shortcut + path[j + 1:]
+        return path
+
+
+class BiRRT(RRT[_RRTState]):
+    """Bidirectional rapidly-exploring random tree."""
+
+    def __init__(self, sample_fn: Callable[[_RRTState], _RRTState],
+                 extend_fn: Callable[[_RRTState, _RRTState],
+                                     Iterator[_RRTState]],
+                 collision_fn: Callable[[_RRTState], bool],
+                 distance_fn: Callable[[_RRTState, _RRTState],
+                                       float], rng: np.random.Generator,
+                 num_attempts: int, num_iters: int, smooth_amt: int):
+        # _sample_goal_eps and _goal_fn aren't used in BiRRT so we just pass
```

Comment:
Wouldn't that force BiRRT usages to pass in a `sample_goal_eps` parameter?

Reply:
good point, I was confused because normally the subclass has more kwargs than the parent class, and this is the other way around. to try to clear up that confusion, what do you think about make `sample_goal_eps` a parameter to the `query` method instead? (going along with my other comment about `query_to_goal_fn`)

Reply:
Sure, that makes sense to me.